pika/adapters/select_connection.py

```diff
@@ -264,20 +263,30 @@ def get_next_deadline(self):
         timeout = min((timeout, SelectPoller.POLL_TIMEOUT))
         return timeout * SelectPoller.POLL_TIMEOUT_MULT
 
+
     def process_timeouts(self):
         """Process the self._timeouts event stack"""
 
         now = time.time()
-        to_run = [timer for timer in self._timeouts.values()
-                  if timer['deadline'] <= now]
-
         # Run the timeouts in order of deadlines. Although this shouldn't
         # be strictly necessary it preserves old behaviour when timeouts
         # were only run periodically.
-        for t in sorted(to_run, key=itemgetter('deadline')):
-            t['callback']()
-            del self._timeouts[hash(frozenset(t.items()))]
-            self._next_timeout = None
+        to_run = sorted([(k, timer) for (k, timer) in self._timeouts.items()
```

Comment:
Unnecessary overhead: no need for the list `[(k, timer) ...`. `sorted` accepts generators/iterators, too. So, this would produce the equivalent result without the unnecessary overhead of instantiating/populating  the list:

```
to_run = sorted(((k, timer) for (k, timer) in self._timeouts.items()
                 if timer['deadline'] <= now),
                key=lambda item: item[1]['deadline'])
```

The original code had this problem, too, so fixing it in your PR is technically out of scope.